themes/tdr/login/src/webauthn/index.ts

```diff
@@ -0,0 +1,266 @@
+import base64url from "base64url"
+
+export default class WebAuthn {
+  setInputValue(id: string, value: string): void {
+    const input: HTMLInputElement | null = document.querySelector(`#${id}`)
+    if (input) {
+      input.value = value
+    }
+  }
+
+  getValueFromInput(id: string): string {
+    const input: HTMLInputElement | null = document.querySelector(`#${id}`)
+    if (input) {
+      return input.value
+    } else {
+      throw `input element with id ${id} is missing`
+    }
+  }
+
+  challenge = this.getValueFromInput("challenge")
+  createTimeout = parseInt(this.getValueFromInput("createTimeout"), 10)
+
+  checkIfWebAuthnSupported() {
+    if (!window.PublicKeyCredential) {
+      const unsupportedMessage = this.getValueFromInput("unsupported-browser-message")
+      this.setInputValue("error", unsupportedMessage)
+      const form: HTMLFormElement | null = document.querySelector("#register")
+      if (form) {
+        form.submit()
+      }
+    }
+  }
+
+  public webAuthnAuthenticate(): void {
+    let isUserIdentified = this.getValueFromInput("isUserIdentified")
+    if (!isUserIdentified) {
+      this.doAuthenticate([])
+      return
+    }
+    this.checkAllowCredentials()
+  }
+
+  checkAllowCredentials() {
+    let allowCredentials: PublicKeyCredentialDescriptor[] = []
+    const useCheckElements: HTMLCollectionOf<Element> = document.getElementsByClassName("authn_use_check")
+    for (const authnUseCheck of useCheckElements) {
+
+      allowCredentials.push({
+        id: base64url.toBuffer((authnUseCheck as HTMLInputElement).value),
+        type: 'public-key',
+      })
+    }
+    this.doAuthenticate(allowCredentials)
+  }
+
```

Comment:
Unnecessary line break

Reply:
This has all changed now so I think this is ok.